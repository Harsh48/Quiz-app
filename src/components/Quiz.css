.quiz-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    background: #fff;
    padding: 20px;
    border-radius: 8px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  }
  
  .timer {
    font-size: 1.2em;
    margin-bottom: 20px;
  }
  
  .start-button {
    background-color: #4CAF50;
    color: white;
    padding: 15px 32px;
    text-align: center;
    font-size: 16px;
    margin: 4px 2px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
  }
  
  .start-button:hover {
    background-color: #45a049;
  }
  