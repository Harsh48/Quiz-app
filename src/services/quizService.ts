import axios from 'axios';

export const fetchQuizQuestions = async (category: string, difficulty: string) => {
  try{
    const response = await axios.get(`https://opentdb.com/api.php?amount=10&category=${category}&difficulty=${difficulty}`);
    return response.data.results;
  }catch(e){
    console.log(e)
  }
};
