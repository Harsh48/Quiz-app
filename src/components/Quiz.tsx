import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { RootState } from '../redux/store';
import { nextQuestion, addScore, finishQuiz, setQuestions, resetQuiz, addAnswer } from '../redux/slices/quizSlice';
import { fetchQuizQuestions } from '../services/quizService';
import Question from './Question';
import Results from './Results';
import CategorySelection from './CategorySelection';
import DifficultySelection from './DifficultySelection';
import GlobalStyles from '../styles/GlobalStyles';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import styled from 'styled-components';

const QuizContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  animation: fadeIn 1s ease-in;
`;

const Timer = styled.div`
  font-size: 1.2em;
  margin-bottom: 20px;
`;

const StartButton = styled.button`
  background-color: #4CAF50;
  color: white;
  padding: 15px 32px;
  font-size: 16px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
`;

const Quiz: React.FC = () => {
  const dispatch = useDispatch();
  const { questions, currentQuestionIndex, status } = useSelector((state: RootState) => state.quiz);
  const [category, setCategory] = useState('');
  const [difficulty, setDifficulty] = useState('');
  const [timeLeft, setTimeLeft] = useState(30);

  useEffect(() => {
    if (timeLeft === 0) {
      handleAnswer(false);
    }
    const timer = setInterval(() => {
      setTimeLeft((prevTime) => prevTime - 1);
    }, 1000);
    return () => clearInterval(timer);
  }, [timeLeft]);

  useEffect(() => {
    console.log('Questions in State:', questions);
  }, [questions]);

  const startQuiz = async () => {
    try {
      const fetchedQuestions = await fetchQuizQuestions(category, difficulty);
      console.log('Fetched Questions:', fetchedQuestions);
      dispatch(setQuestions(fetchedQuestions));
    } catch (error) {
      toast.error('Error fetching questions. Please try again.');
      console.error('Error fetching questions:', error);
    }
  };

  const handleAnswer = (isCorrect: boolean) => {
    const question = questions[currentQuestionIndex];
    if(question){
    dispatch(addAnswer({
      question: question?.question,
      answer: isCorrect ? question.correct_answer : question.incorrect_answers[0], // Simplification for demo purposes
      correct: isCorrect,
    }));
   }
    if (isCorrect) {
      dispatch(addScore(questions[currentQuestionIndex].type === 'multiple' ? 10 : 5));
    }

    if (currentQuestionIndex < questions.length - 1) {
      dispatch(nextQuestion());
      setTimeLeft(30);
    } else {
      dispatch(finishQuiz());
    }
  };

  const handleReset = () => {
    dispatch(resetQuiz());
    setTimeLeft(30);
  };

  if (status === 'idle') {
    return (
      <QuizContainer>
        <GlobalStyles />
        <CategorySelection setCategory={setCategory} />
        <DifficultySelection setDifficulty={setDifficulty} />
        <StartButton onClick={startQuiz}>Start Quiz</StartButton>
        <ToastContainer />
      </QuizContainer>
    );
  }

  if (status === 'finished') {
    return <Results onReset={handleReset} />;
  }

  return (
    <QuizContainer>
      <GlobalStyles />
      <Timer>Time left: {timeLeft} seconds</Timer>
      {questions.length > 0 && (
        <Question question={questions[currentQuestionIndex]} onAnswer={handleAnswer} />
      )}
      <ToastContainer />
    </QuizContainer>
  );
};

export default Quiz;
