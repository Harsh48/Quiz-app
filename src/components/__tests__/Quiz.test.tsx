import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Quiz from '../Quiz';
import { RootState } from '../../redux/store';
import 'jest-styled-components';
import { fetchQuizQuestions } from '../../services/quizService';

jest.mock('../../services/quizService');

const mockStore = configureStore([]);
const mockedFetchQuizQuestions = fetchQuizQuestions as jest.MockedFunction<typeof fetchQuizQuestions>;

const initialState: RootState = {
  quiz: {
    questions: [
      {
        category: 'General Knowledge',
        type: 'multiple',
        difficulty: 'easy',
        question: 'What is the capital of France?',
        correct_answer: 'Paris',
        incorrect_answers: ['London', 'Berlin', 'Madrid'],
      },
    ],
    currentQuestionIndex: 0,
    score: 0,
    status: 'idle',
    answers: [],
  },
};

test('renders Quiz component correctly', () => {
  const store = mockStore(initialState);
  const { asFragment } = render(
    <Provider store={store}>
      <Quiz />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('starts quiz on button click', async () => {
  const store = mockStore(initialState);
  mockedFetchQuizQuestions.mockResolvedValueOnce(initialState.quiz.questions);

  render(
    <Provider store={store}>
      <Quiz />
    </Provider>
  );

  fireEvent.click(screen.getByText('Start Quiz'));

  await waitFor(() => {
    const actions = store.getActions();
    expect(actions).toContainEqual({ type: 'quiz/setQuestions', payload: initialState.quiz.questions });
  });
});

test('displays question and handles answer selection', () => {
  const stateWithActiveQuiz: RootState = {
    ...initialState,
    quiz: { ...initialState.quiz, status: 'active' }
  };

  const store = mockStore(stateWithActiveQuiz);
  render(
    <Provider store={store}>
      <Quiz />
    </Provider>
  );

  expect(screen.getByText('What is the capital of France?')).toBeInTheDocument();
  fireEvent.click(screen.getByText('Paris'));
  const actions = store.getActions();
  expect(actions).toContainEqual({ type: 'quiz/addScore', payload: 10 });
  expect(actions).toContainEqual({
    type: 'quiz/addAnswer',
    payload: { question: 'What is the capital of France?', answer: 'Paris', correct: true },
  });
});
