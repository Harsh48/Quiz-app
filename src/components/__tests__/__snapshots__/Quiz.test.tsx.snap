// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Quiz component correctly 1`] = `
<DocumentFragment>
  .c1 {
  margin-bottom: 20px;
  animation: fadeIn 1s ease-in;
}

.c2 {
  padding: 10px;
  font-size: 16px;
  border-radius: 4px;
  border: 1px solid #ccc;
}

.c3 {
  margin-bottom: 20px;
  animation: fadeIn 1s ease-in;
}

.c4 {
  padding: 10px;
  font-size: 16px;
  border-radius: 4px;
  border: 1px solid #ccc;
}

.c0 {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #fff;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  animation: fadeIn 1s ease-in;
}

.c5 {
  background-color: #4CAF50;
  color: white;
  padding: 15px 32px;
  font-size: 16px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h3>
        Select Category:
      </h3>
      <select
        class="c2"
      >
        <option
          value="9"
        >
          General Knowledge
        </option>
        <option
          value="10"
        >
          Entertainment: Books
        </option>
      </select>
    </div>
    <div
      class="c3"
    >
      <h3>
        Select Difficulty:
      </h3>
      <select
        class="c4"
      >
        <option
          value="easy"
        >
          easy
        </option>
        <option
          value="medium"
        >
          medium
        </option>
        <option
          value="hard"
        >
          hard
        </option>
      </select>
    </div>
    <button
      class="c5"
    >
      Start Quiz
    </button>
    <div
      class="Toastify"
    />
  </div>
</DocumentFragment>
`;
